Mathlib/Combinatorics/Quiver/Path/Vertices.lean

```diff
@@ -160,7 +160,201 @@ lemma comp_eq_nil_iff {a b : V} {p : Path a b} {q : Path b a} :
+    have : l₁ ++ v :: l₂ = p₁.vertices.dropLast ++ p₂.vertices := by
+      rw [← hv, h_verts]
+    exact append_inj_left this h_l1_len
+  have h_v_l2 : v :: l₂ = p₂.vertices := by
+    have : l₁ ++ v :: l₂ = p₁.vertices.dropLast ++ p₂.vertices := by
+      rw [← hv, h_verts]
+    rw [h_l1_eq] at this
+    exact append_cancel_left this
+  have : p₂.vertices.head? = some c := by
+    cases p₂ with
+    | nil => simp only [vertices_nil, head?_cons]
+    | cons _ _ => exact vertices_head? _
+  rw [← h_v_l2] at this
+  simp [head?_cons, Option.some.injEq] at this
+  exact this
+
+/-- `split_at_vertex` decomposes a path `p` at the vertex sitting in
+    position `i` of its `vertices` -/
+theorem split_at_vertices_index (i : ℕ)
+    (hi : i < p.vertices.length) :
+    ∃ (v : V) (p₁ : Path a v) (p₂ : Path v b),
+      p = p₁.comp p₂ ∧
+      p₁.length = i ∧
+      v = p.vertices[i] := by
+  have hi_le_len : i ≤ p.length := by
+    rw [vertices_length] at hi
+    exact Nat.le_of_lt_succ hi
+  obtain ⟨v, p₁, p₂, rfl, rfl⟩ := exists_eq_comp_of_le_length p hi_le_len
+  exact ⟨v, p₁, p₂, rfl, rfl, by simp⟩
+
+/-- Given vertices lists from a path composition,
+the prefix path’s vertices is a prefix of the full path’s vertices. -/
+lemma vertices_dropLast_isPrefix_of_vertices_eq
+    {V : Type*} [Quiver V]
+    {a b c : V} {p : Path a b} {p₁ : Path a c} {p₂ : Path c b}
+    (h : p.vertices = p₁.vertices.dropLast ++ p₂.vertices) :
+    p₁.vertices.dropLast.IsPrefix p.vertices := by
+  rw [h]; exact prefix_append p₁.vertices.dropLast p₂.vertices
+
+lemma mem_vertices_eq_end_or_mem_dropLast {a b c : V} (p : Path a b) (h : c ∈ p.vertices) :
+  c = b ∨ c ∈ p.vertices.dropLast := by
+  induction p with
+  | nil =>
+    simp [vertices_nil] at h
+    subst h
+    simp
+  | cons p' e ih =>
+    rename_i b'
+    have h' : c ∈ p'.vertices ∨ c = b' := by
+      simpa using (mem_vertices_cons p' e).1 h
+    cases h' with
+    | inl h_in_p' =>
+      have ih' := ih h_in_p'
+      cases ih' with
+      | inl h_eq_p'_end =>
+        right
+        subst h_eq_p'_end
+        simp
+      | inr h_in_p'_dropLast =>
+        simp_all
+    | inr h_eq_b =>
+      left
+      exact h_eq_b
+
+/-- If we have a path p from a to b with c ∈ p.vertices,
+    and c is not the end vertex b, then it appears in a proper prefix of the path. -/
+lemma exists_prefix_of_mem_vertices_ne_end [DecidableEq V] {a b c : V}
+    (p : Path a b) (h : c ∈ p.vertices) (h_ne : c ≠ b) :
+  ∃ (p₁ : Path a c) (p₂ : Path c b), p = p₁.comp p₂ := by
+  have h_cases := mem_vertices_eq_end_or_mem_dropLast p h
+  cases h_cases with
+  | inl h_eq =>
+      contradiction
+  | inr h_mem_tail =>
+      let i := p.vertices.idxOf c
+      have hi : i < p.vertices.length := List.idxOf_lt_length_iff.2 h
+      obtain ⟨v, p₁, p₂, h_comp, h_len, h_c_eq⟩ := split_at_vertices_index p i hi
+      have hvc : v = c := by
+        rw [h_c_eq]
+        simp [i]
+      subst hvc
+      exact ⟨p₁, p₂, h_comp⟩
+
```

Comment:
Let's just drop this lemma: the hypothesis `h_ne` is redundant (to witness this, replace the `contradiction` tactic by just `subst h_eq; exact ⟨p, Path.nil, by simp⟩`) and it is just a weaker form of the lemma immediately below.
```suggestion
```